README.md
## Overview

Upgrade Container for Gluu Server  Community Conatiner Edition and Enterprise Edition.

This will be run as a job in Kubernetes or separate container in Docker. It will automatically upgrade your Gluu Server static configuration and LDAP data to the labeled version.

## Versions

- Stable: `gluufederation/upgrade:4.2.0_01`
- Unstable: `gluufederation/upgrade:4.2.1_dev`

Refer to [Changelog](./CHANGES.md) for details on new features, bug fixes, or older releases.

## Environment Variables

The following environment variables are supported by the container:

- `GLUU_CONFIG_ADAPTER`: The config backend adapter, can be `consul` (default) or `kubernetes`.
- `GLUU_CONFIG_CONSUL_HOST`: hostname or IP of Consul (default to `localhost`).
- `GLUU_CONFIG_CONSUL_PORT`: port of Consul (default to `8500`).
- `GLUU_CONFIG_CONSUL_CONSISTENCY`: Consul consistency mode (choose one of `default`, `consistent`, or `stale`). Default to `stale` mode.
- `GLUU_CONFIG_CONSUL_SCHEME`: supported Consul scheme (`http` or `https`).
- `GLUU_CONFIG_CONSUL_VERIFY`: whether to verify cert or not (default to `false`).
- `GLUU_CONFIG_CONSUL_CACERT_FILE`: path to Consul CA cert file (default to `/etc/certs/consul_ca.crt`). This file will be used if it exists and `GLUU_CONFIG_CONSUL_VERIFY` set to `true`.
- `GLUU_CONFIG_CONSUL_CERT_FILE`: path to Consul cert file (default to `/etc/certs/consul_client.crt`).
- `GLUU_CONFIG_CONSUL_KEY_FILE`: path to Consul key file (default to `/etc/certs/consul_client.key`).
- `GLUU_CONFIG_CONSUL_TOKEN_FILE`: path to file contains ACL token (default to `/etc/certs/consul_token`).
- `GLUU_CONFIG_KUBERNETES_NAMESPACE`: Kubernetes namespace (default to `default`).
- `GLUU_CONFIG_KUBERNETES_CONFIGMAP`: Kubernetes configmaps name (default to `gluu`).
- `GLUU_CONFIG_KUBERNETES_USE_KUBE_CONFIG`: Load credentials from `$HOME/.kube/config`, only useful for non-container environment (default to `false`).
- `GLUU_SECRET_ADAPTER`: The secrets adapter, can be `vault` or `kubernetes`.
- `GLUU_SECRET_VAULT_SCHEME`: supported Vault scheme (`http` or `https`).
- `GLUU_SECRET_VAULT_HOST`: hostname or IP of Vault (default to `localhost`).
- `GLUU_SECRET_VAULT_PORT`: port of Vault (default to `8200`).
- `GLUU_SECRET_VAULT_VERIFY`: whether to verify cert or not (default to `false`).
- `GLUU_SECRET_VAULT_ROLE_ID_FILE`: path to file contains Vault AppRole role ID (default to `/etc/certs/vault_role_id`).
- `GLUU_SECRET_VAULT_SECRET_ID_FILE`: path to file contains Vault AppRole secret ID (default to `/etc/certs/vault_secret_id`).
- `GLUU_SECRET_VAULT_CERT_FILE`: path to Vault cert file (default to `/etc/certs/vault_client.crt`).
- `GLUU_SECRET_VAULT_KEY_FILE`: path to Vault key file (default to `/etc/certs/vault_client.key`).
- `GLUU_SECRET_VAULT_CACERT_FILE`: path to Vault CA cert file (default to `/etc/certs/vault_ca.crt`). This file will be used if it exists and `GLUU_SECRET_VAULT_VERIFY` set to `true`.
- `GLUU_SECRET_KUBERNETES_NAMESPACE`: Kubernetes namespace (default to `default`).
- `GLUU_SECRET_KUBERNETES_CONFIGMAP`: Kubernetes secrets name (default to `gluu`).
- `GLUU_SECRET_KUBERNETES_USE_KUBE_CONFIG`: Load credentials from `$HOME/.kube/config`, only useful for non-container environment (default to `false`).
- `GLUU_WAIT_MAX_TIME`: How long the startup "health checks" should run (default to `300` seconds).
- `GLUU_WAIT_SLEEP_DURATION`: Delay between startup "health checks" (default to `10` seconds).
- `GLUU_PERSISTENCE_TYPE`: Persistence backend being used (one of `ldap`, `couchbase`, or `hybrid`; default to `ldap`).
- `GLUU_PERSISTENCE_LDAP_MAPPING`: Specify data that should be saved in LDAP (one of `default`, `user`, `cache`, `site`, or `token`; default to `default`). Note this environment only takes effect when `GLUU_PERSISTENCE_TYPE` is set to `hybrid`.
- `GLUU_LDAP_URL`: Address and port of LDAP server (default to `localhost:1636`); required if `GLUU_PERSISTENCE_TYPE` is set to `ldap` or `hybrid`.
- `GLUU_COUCHBASE_URL`: Address of Couchbase server (default to `localhost`); required if `GLUU_PERSISTENCE_TYPE` is set to `couchbase` or `hybrid`.
- `GLUU_COUCHBASE_USER`: Username of Couchbase server (default to `admin`); required if `GLUU_PERSISTENCE_TYPE` is set to `couchbase` or `hybrid`.
- `GLUU_COUCHBASE_CERT_FILE`: Couchbase root certificate location (default to `/etc/certs/couchbase.crt`); required if `GLUU_PERSISTENCE_TYPE` is set to `couchbase` or `hybrid`.
- `GLUU_COUCHBASE_PASSWORD_FILE`: Path to file contains Couchbase password (default to `/etc/gluu/conf/couchbase_password`); required if `GLUU_PERSISTENCE_TYPE` is set to `couchbase` or `hybrid`.

## Entrypoint Parameters

The following parameters are supported by the container:

- `--source`: Gluu version to upgrade from (required)
- `--target`: Gluu version to upgrade to; usually the latest version (required)
- `--help`: Show all available parameters

**Docker:**
   ```bash
    docker run \
        --rm \
        --network container:consul \
        -e GLUU_CONFIG_CONSUL_HOST=consul \
        -e GLUU_SECRET_VAULT_HOST=vault \
        -e GLUU_PERSISTENCE_TYPE=ldap \
        -e GLUU_LDAP_URL=ldap:1636 \
        -v /path/to/vault_role_id.txt:/etc/certs/vault_role_id \
        -v /path/to/vault_secret_id.txt:/etc/certs/vault_secret_id \
        gluufederation/upgrade:4.2.0_01 \
            --source 4.1 \
            --target 4.2
   ```

**Kubernetes:**

   ```yaml
    apiVersion: batch/v1
    kind: Job
    metadata:
      name: gluu-upgrade-job
    spec:
      template:
        spec:
          restartPolicy: Never
          containers:
            - name: gluu-upgrade-job
              image: gluufederation/upgrade:4.2.0_01
              envFrom:
              - configMapRef:
                  name: upgrade-cm
              args: ["--source", "4.1", "--target", "4.2"]

   ```


CHANGES.md
# Changelog

Here you can see an overview of changes between each release.

## Version 4.2.0_01

Released on July 20th, 2020.

* Added support for Gluu Server v4.2.
* Upraded to Java 11.
* Upgraded to Python3.

## Version 4.1.1_04

Released on June 5th, 2020.

* Upgraded `oxauth-client`.
* Added feature to update picture attribute.

## Version 4.1.1_03

Released on April 29th, 2020.

* Added DocumentStore support.

## Version 4.1.1_02

Released on April 2nd, 2020.

* Upgraded `oxauth-client`.

## Version 4.1.1_01

Released on March 24th, 2020.

* Conformed to Gluu Server v4.1.1.

## Version 4.1.0_01

Released on March 5th, 2020.

* Conformed to Gluu Server v4.1.

## Version 4.0.1_01

Released on November 1st, 2019.

* Upgraded to Gluu Server 4.0.1.
* Removed `4.0.0` target version in favor of `4.0` version.

## Version 4.0.0_01

Released on October 22nd, 2019.

* Added feature to upgrade Gluu Server v3 to v4.

## Version 3.1.6_03

Released on July 4th, 2019.

* Added missing `shareSubjectIdBetweenClientsWithSameSectorId` config for oxAuth. Ref: https://github.com/GluuFederation/gluu-docker/issues/78.

## Version 3.1.6_02

Released on May 10th, 2019.

* Alpine upgraded to v3.9. Ref: https://github.com/GluuFederation/gluu-docker/issues/71.

## Version 3.1.6_01

Released on April 29th, 2019.

* Upgraded to Gluu Server 3.1.6.

## Version 3.1.5_04

Released on May 10th, 2019.

* Alpine upgraded to v3.9. Ref: https://github.com/GluuFederation/gluu-docker/issues/71.

## Version 3.1.5_03

Released on April 9th, 2019.

* Added license info on container startup.

## Version 3.1.5_02

Released on March 27th, 2019.

* Changed license to GLUU-SUPPORT.

## Version 3.1.5_01

Released on March 23rd, 2019.

* Add feature to upgrade data from Gluu Server 3.1.4 to 3.1.5.


scripts/gluulib.py
import base64
import json
import logging
import os
from collections import namedtuple

import hvac
import six
import kubernetes.client
import kubernetes.config
from consul import Consul

logger = logging.getLogger("gluulib")
logger.setLevel(logging.INFO)
ch = logging.StreamHandler()
fmt = logging.Formatter('%(levelname)s - %(asctime)s - %(message)s')
ch.setFormatter(fmt)
logger.addHandler(ch)


def as_boolean(val, default=False):
    """Converts value as boolean.
    """
    truthy = set(('t', 'T', 'true', 'True', 'TRUE', '1', 1, True))
    falsy = set(('f', 'F', 'false', 'False', 'FALSE', '0', 0, False))

    if val in truthy:
        return True
    if val in falsy:
        return False
    return default


def safe_value(value):
    if not isinstance(value, (six.string_types, six.binary_type)):
        value = json.dumps(value)
    return value


class BaseConfig(object):
    """Base class for config adapter. Must be sub-classed per
    implementation details.
    """
    type = "config"

    def get(self, key, default=None):
        """Get specific config.

        Subclass __MUST__ implement this method.
        """
        raise NotImplementedError

    def set(self, key, value):
        """Set specific config.

        Subclass __MUST__ implement this method.
        """
        raise NotImplementedError

    def all(self):
        """Get all config as ``dict`` type.

        Subclass __MUST__ implement this method.
        """
        raise NotImplementedError


class ConsulConfig(BaseConfig):
    def __init__(self):
        # collects all env vars prefixed with `GLUU_CONFIG_CONSUL_`,
        # for example `GLUU_CONFIG_CONSUL_HOST=localhost`
        self.settings = {
            k: v for k, v in os.environ.iteritems()
            if k.isupper() and k.startswith("GLUU_CONFIG_CONSUL_")
        }

        self.settings.setdefault(
            "GLUU_CONFIG_CONSUL_HOST",
            # backward-compat with Gluu Server v3.1.4
            os.environ.get("GLUU_CONSUL_HOST", "localhost"),
        )

        self.settings.setdefault(
            "GLUU_CONFIG_CONSUL_PORT",
            # backward-compat with Gluu Server v3.1.4
            os.environ.get("GLUU_CONSUL_PORT", 8500),
        )

        self.settings.setdefault(
            "GLUU_CONFIG_CONSUL_CONSISTENCY",
            # backward-compat with Gluu Server v3.1.4
            os.environ.get("GLUU_CONSUL_CONSISTENCY", "stale"),
        )

        self.settings.setdefault(
            "GLUU_CONFIG_CONSUL_SCHEME",
            # backward-compat with Gluu Server v3.1.4
            os.environ.get("GLUU_CONSUL_SCHEME", "http"),
        )

        self.settings.setdefault(
            "GLUU_CONFIG_CONSUL_VERIFY",
            # backward-compat with Gluu Server v3.1.4
            os.environ.get("GLUU_CONSUL_VERIFY", False),
        )

        self.settings.setdefault(
            "GLUU_CONFIG_CONSUL_CACERT_FILE",
            # backward-compat with Gluu Server v3.1.4
            os.environ.get("GLUU_CONSUL_CACERT_FILE",
                           "/etc/certs/consul_ca.crt"),
        )

        self.settings.setdefault(
            "GLUU_CONFIG_CONSUL_CERT_FILE",
            # backward-compat with Gluu Server v3.1.4
            os.environ.get("GLUU_CONSUL_CERT_FILE",
                           "/etc/certs/consul_client.crt"),
        )

        self.settings.setdefault(
            "GLUU_CONFIG_CONSUL_KEY_FILE",
            # backward-compat with Gluu Server v3.1.4
            os.environ.get("GLUU_CONSUL_KEY_FILE",
                           "/etc/certs/consul_client.key"),
        )

        self.settings.setdefault(
            "GLUU_CONFIG_CONSUL_TOKEN_FILE",
            # backward-compat with Gluu Server v3.1.4
            os.environ.get("GLUU_CONSUL_TOKEN_FILE", "/etc/certs/consul_token"),
        )

        self.prefix = "gluu/config/"
        token = None
        cert = None
        verify = False

        if os.path.isfile(self.settings["GLUU_CONFIG_CONSUL_TOKEN_FILE"]):
            with open(self.settings["GLUU_CONFIG_CONSUL_TOKEN_FILE"]) as fr:
                token = fr.read().strip()

        if self.settings["GLUU_CONFIG_CONSUL_SCHEME"] == "https":
            verify = as_boolean(self.settings["GLUU_CONFIG_CONSUL_VERIFY"])

            # verify using CA cert (if any)
            if all([verify,
                    os.path.isfile(self.settings["GLUU_CONFIG_CONSUL_CACERT_FILE"])]):
                verify = self.settings["GLUU_CONFIG_CONSUL_CACERT_FILE"]

            if all([os.path.isfile(self.settings["GLUU_CONFIG_CONSUL_CERT_FILE"]),
                    os.path.isfile(self.settings["GLUU_CONFIG_CONSUL_KEY_FILE"])]):
                cert = (self.settings["GLUU_CONFIG_CONSUL_CERT_FILE"],
                        self.settings["GLUU_CONFIG_CONSUL_KEY_FILE"])

        self._request_warning(self.settings["GLUU_CONFIG_CONSUL_SCHEME"], verify)

        self.client = Consul(
            host=self.settings["GLUU_CONFIG_CONSUL_HOST"],
            port=self.settings["GLUU_CONFIG_CONSUL_PORT"],
            token=token,
            scheme=self.settings["GLUU_CONFIG_CONSUL_SCHEME"],
            consistency=self.settings["GLUU_CONFIG_CONSUL_CONSISTENCY"],
            verify=verify,
            cert=cert,
        )

    def _merge_path(self, key):
        """Add prefix to the key.
        """
        return "".join([self.prefix, key])

    def _unmerge_path(self, key):
        """Remove prefix from the key.
        """
        return key[len(self.prefix):]

    def get(self, key, default=None):
        _, result = self.client.kv.get(self._merge_path(key))
        if not result:
            return default
        return result["Value"]

    def set(self, key, value):
        return self.client.kv.put(self._merge_path(key),
                                  safe_value(value))

    def find(self, key):
        _, resultset = self.client.kv.get(self._merge_path(key),
                                          recurse=True)

        if not resultset:
            return {}

        return {
            self._unmerge_path(item["Key"]): item["Value"]
            for item in resultset
        }

    def all(self):
        return self.find("")

    def _request_warning(self, scheme, verify):
        if scheme == "https" and verify is False:
            import urllib3
            urllib3.disable_warnings()
            logger.warn(
                "All requests to Consul will be unverified. "
                "Please adjust GLUU_CONFIG_CONSUL_SCHEME and "
                "GLUU_CONFIG_CONSUL_VERIFY environment variables."
            )


class KubernetesConfig(BaseConfig):
    def __init__(self):
        self.settings = {
            k: v for k, v in os.environ.iteritems()
            if k.isupper() and k.startswith("GLUU_CONFIG_KUBERNETES_")
        }
        self.settings.setdefault(
            "GLUU_CONFIG_KUBERNETES_NAMESPACE",
            # backward-compat with Gluu Server v3.1.4
            os.environ.get("GLUU_KUBERNETES_NAMESPACE", "default"),
        )

        self.settings.setdefault(
            "GLUU_CONFIG_KUBERNETES_CONFIGMAP",
            # backward-compat with Gluu Server v3.1.4
            os.environ.get("GLUU_KUBERNETES_CONFIGMAP", "gluu"),
        )

        self.settings.setdefault(
            "GLUU_CONFIG_KUBERNETES_USE_KUBE_CONFIG",
            False
        )

        if as_boolean(self.settings["GLUU_CONFIG_KUBERNETES_USE_KUBE_CONFIG"]):
            kubernetes.config.load_kube_config()
        else:
            kubernetes.config.load_incluster_config()

        self.client = kubernetes.client.CoreV1Api()
        self.name_exists = False

    def get(self, key, default=None):
        result = self.all()
        return result.get(key, default)

    def _prepare_configmap(self):
        # create a configmap name if not exist
        if not self.name_exists:
            try:
                self.client.read_namespaced_config_map(
                    self.settings["GLUU_CONFIG_KUBERNETES_CONFIGMAP"],
                    self.settings["GLUU_CONFIG_KUBERNETES_NAMESPACE"])
                self.name_exists = True
            except kubernetes.client.rest.ApiException as exc:
                if exc.status == 404:
                    # create the configmaps name
                    body = {
                        "kind": "ConfigMap",
                        "apiVersion": "v1",
                        "metadata": {
                            "name": self.settings["GLUU_CONFIG_KUBERNETES_CONFIGMAP"],
                        },
                        "data": {},
                    }
                    created = self.client.create_namespaced_config_map(
                        self.settings["GLUU_CONFIG_KUBERNETES_NAMESPACE"],
                        body)
                    if created:
                        self.name_exists = True
                else:
                    raise

    def set(self, key, value):
        self._prepare_configmap()
        body = {
            "kind": "ConfigMap",
            "apiVersion": "v1",
            "metadata": {
                "name": self.settings["GLUU_CONFIG_KUBERNETES_CONFIGMAP"],
            },
            "data": {
                key: safe_value(value),
            }
        }
        return self.client.patch_namespaced_config_map(
            self.settings["GLUU_CONFIG_KUBERNETES_CONFIGMAP"],
            self.settings["GLUU_CONFIG_KUBERNETES_NAMESPACE"],
            body=body)

    def all(self):
        self._prepare_configmap()
        result = self.client.read_namespaced_config_map(
            self.settings["GLUU_CONFIG_KUBERNETES_CONFIGMAP"],
            self.settings["GLUU_CONFIG_KUBERNETES_NAMESPACE"])
        return result.data or {}


class ConfigManager(object):
    def __init__(self):
        _adapter = os.environ.get(
            "GLUU_CONFIG_ADAPTER",
            "consul",
        )
        if _adapter == "consul":
            self.adapter = ConsulConfig()
        elif _adapter == "kubernetes":
            self.adapter = KubernetesConfig()
        else:
            self.adapter = None

    def get(self, key, default=None):
        return self.adapter.get(key, default)

    def set(self, key, value):
        return self.adapter.set(key, value)

    def all(self):
        return self.adapter.all()


class BaseSecret(object):
    """Base class for secret adapter. Must be sub-classed per
    implementation details.
    """
    type = "secret"

    def get(self, key, default=None):
        """Get specific secret.

        Subclass __MUST__ implement this method.
        """
        raise NotImplementedError

    def set(self, key, value):
        """Set specific secret.

        Subclass __MUST__ implement this method.
        """
        raise NotImplementedError

    def all(self):
        """Get all secrets as ``dict`` type.

        Subclass __MUST__ implement this method.
        """
        raise NotImplementedError


class VaultSecret(BaseSecret):
    def __init__(self):
        self.settings = {
            k: v for k, v in os.environ.iteritems()
            if k.isupper() and k.startswith("GLUU_SECRET_VAULT_")
        }
        self.settings.setdefault(
            "GLUU_SECRET_VAULT_HOST",
            "localhost",
        )
        self.settings.setdefault(
            "GLUU_SECRET_VAULT_PORT",
            8200,
        )
        self.settings.setdefault(
            "GLUU_SECRET_VAULT_SCHEME",
            "http",
        )
        self.settings.setdefault(
            "GLUU_SECRET_VAULT_VERIFY",
            False,
        )
        self.settings.setdefault(
            "GLUU_SECRET_VAULT_ROLE_ID_FILE",
            "/etc/certs/vault_role_id",
        ),
        self.settings.setdefault(
            "GLUU_SECRET_VAULT_SECRET_ID_FILE",
            "/etc/certs/vault_secret_id",
        )
        self.settings.setdefault(
            "GLUU_SECRET_VAULT_CERT_FILE",
            "/etc/certs/vault_client.crt",
        )
        self.settings.setdefault(
            "GLUU_SECRET_VAULT_KEY_FILE",
            "/etc/certs/vault_client.key",
        )
        self.settings.setdefault(
            "GLUU_SECRET_VAULT_CACERT_FILE",
            "/etc/certs/vault_ca.crt",
        )

        cert = None
        verify = False

        if self.settings["GLUU_SECRET_VAULT_SCHEME"] == "https":
            verify = as_boolean(self.settings["GLUU_SECRET_VAULT_VERIFY"])

            # verify using CA cert (if any)
            if all([verify,
                    os.path.isfile(self.settings["GLUU_SECRET_VAULT_CACERT_FILE"])]):
                verify = self.settings["GLUU_SECRET_VAULT_CACERT_FILE"]

            if all([os.path.isfile(self.settings["GLUU_SECRET_VAULT_CERT_FILE"]),
                    os.path.isfile(self.settings["GLUU_SECRET_VAULT_KEY_FILE"])]):
                cert = (self.settings["GLUU_SECRET_VAULT_CERT_FILE"],
                        self.settings["GLUU_SECRET_VAULT_KEY_FILE"])

        self._request_warning(self.settings["GLUU_SECRET_VAULT_SCHEME"], verify)

        self.client = hvac.Client(
            url="{}://{}:{}".format(
                self.settings["GLUU_SECRET_VAULT_SCHEME"],
                self.settings["GLUU_SECRET_VAULT_HOST"],
                self.settings["GLUU_SECRET_VAULT_PORT"],
            ),
            cert=cert,
            verify=verify,
        )
        self.prefix = "secret/gluu"

    @property
    def role_id(self):
        try:
            with open(self.settings["GLUU_SECRET_VAULT_ROLE_ID_FILE"]) as f:
                role_id = f.read()
        except IOError:
            role_id = ""
        return role_id

    @property
    def secret_id(self):
        try:
            with open(self.settings["GLUU_SECRET_VAULT_SECRET_ID_FILE"]) as f:
                secret_id = f.read()
        except IOError:
            secret_id = ""
        return secret_id

    def _authenticate(self):
        if self.client.is_authenticated():
            return

        creds = self.client.auth_approle(self.role_id, self.secret_id, use_token=False)
        self.client.token = creds["auth"]["client_token"]

    def get(self, key, default=None):
        self._authenticate()
        sc = self.client.read("{}/{}".format(self.prefix, key))
        if not sc:
            return default
        return sc["data"]["value"]

    def set(self, key, value):
        self._authenticate()
        val = {"value": value}

        # hvac.v1.Client.write checks for status code 200,
        # but Vault HTTP API returns 204 if request succeeded;
        # hence we're using lower level of `hvac.v1.Client` API to set key-val
        response = self.client._adapter.post('/v1/{0}/{1}'.format(self.prefix, key), json=val)
        return response.status_code == 204

    def all(self):
        self._authenticate()
        result = self.client.list(self.prefix)
        return {key: self.get(key) for key in result["data"]["keys"]}

    def _request_warning(self, scheme, verify):
        if scheme == "https" and verify is False:
            import urllib3
            urllib3.disable_warnings()
            logger.warn(
                "All requests to Vault will be unverified. "
                "Please adjust GLUU_SECRET_VAULT_SCHEME and "
                "GLUU_SECRET_VAULT_VERIFY environment variables."
            )


class KubernetesSecret(BaseSecret):
    def __init__(self):
        self.settings = {
            k: v for k, v in os.environ.iteritems()
            if k.isupper() and k.startswith("GLUU_SECRET_KUBERNETES_")
        }
        self.settings.setdefault(
            "GLUU_SECRET_KUBERNETES_NAMESPACE",
            "default",
        )
        self.settings.setdefault(
            "GLUU_SECRET_KUBERNETES_SECRET",
            "gluu",
        )
        self.settings.setdefault(
            "GLUU_SECRET_KUBERNETES_USE_KUBE_CONFIG",
            False
        )

        if as_boolean(self.settings["GLUU_SECRET_KUBERNETES_USE_KUBE_CONFIG"]):
            kubernetes.config.load_kube_config()
        else:
            kubernetes.config.load_incluster_config()
        self.client = kubernetes.client.CoreV1Api()
        self.name_exists = False

    def get(self, key, default=None):
        result = self.all()
        if key in result:
            return base64.b64decode(result[key])
        return default

    def _prepare_secret(self):
        # create a secret name if not exist
        if not self.name_exists:
            try:
                self.client.read_namespaced_secret(
                    self.settings["GLUU_SECRET_KUBERNETES_SECRET"],
                    self.settings["GLUU_SECRET_KUBERNETES_NAMESPACE"])
                self.name_exists = True
            except kubernetes.client.rest.ApiException as exc:
                if exc.status == 404:
                    # create the secrets name
                    body = {
                        "kind": "Secret",
                        "apiVersion": "v1",
                        "metadata": {
                            "name": self.settings["GLUU_SECRET_KUBERNETES_SECRET"],
                        },
                        "data": {},
                    }
                    created = self.client.create_namespaced_secret(
                        self.settings["GLUU_SECRET_KUBERNETES_NAMESPACE"],
                        body)
                    if created:
                        self.name_exists = True
                else:
                    raise

    def set(self, key, value):
        self._prepare_secret()
        body = {
            "kind": "Secret",
            "apiVersion": "v1",
            "metadata": {
                "name": self.settings["GLUU_SECRET_KUBERNETES_SECRET"],
            },
            "data": {
                key: base64.b64encode(value),
            }
        }
        return self.client.patch_namespaced_secret(
            self.settings["GLUU_SECRET_KUBERNETES_SECRET"],
            self.settings["GLUU_SECRET_KUBERNETES_NAMESPACE"],
            body=body)

    def all(self):
        self._prepare_secret()
        result = self.client.read_namespaced_secret(
            self.settings["GLUU_SECRET_KUBERNETES_SECRET"],
            self.settings["GLUU_SECRET_KUBERNETES_NAMESPACE"])
        return result.data or {}


class SecretManager(object):
    def __init__(self, config_adapter=None):
        _adapter = os.environ.get(
            "GLUU_SECRET_ADAPTER",
            "vault",
        )
        if _adapter == "vault":
            self.adapter = VaultSecret()
        elif _adapter == "kubernetes":
            self.adapter = KubernetesSecret()
        else:
            self.adapter = None

        # backward-compat
        self.config_adapter = config_adapter

    def _get_compat(self, key, default=None):
        sc = self.adapter.get(key, default)

        if not sc and self.config_adapter:
            # tries to get from old config
            sc = self.config_adapter.get(key, default)

        if not sc:
            # fallback to default
            sc = default
        return sc

    def get(self, key, default=None):
        return self._get_compat(key, default)

    def set(self, key, value):
        return self.adapter.set(key, value)

    def all(self):
        return self.adapter.all()


def get_manager():
    """Convenient function to get manager instances.
    """
    obj = namedtuple("Manager", "config secret")
    config_mgr = ConfigManager()
    secret_mgr = SecretManager(config_mgr)
    return obj(config=config_mgr, secret=secret_mgr)
